# Άσκηση 3: Βελτιστοποίηση συναρτήσεων με Γενετικούς Αλγόριθμους
v1.0, 08/01/2019

<img src="http://infinity77.net/global_optimization/_images/Alpine01.png" alt="Alpine01" style="width: 500px;"/>

Στόχος της άσκησης είναι η βελτιστοποίηση συναρτήσεων χωρίς παραγώγους (derivative free optimization) με χρήση Γενετικών Αλγόριθμων μέσω της βιβλιοθήκης DEAP. Η βελτιστοποίηση χωρίς παραγώγους είναι ιδιαίτερα χρήσιμη σε περιπτώσεις όπου η αντικειμενική συνάρτηση $f$ δεν είναι διαθέσιμη (black-box optimization) ή σε περίπτωσεις που είναι, είναι δύσκολο ή μη πρακτικό να υπολογιστούν οι παράγωγοί της. Για παράδειγμα η  $f$ μπορεί να μην είναι διαφορίσιμη, η παραγώγιση της μπορεί να είναι δύσκολη ή να απαιτεί πολύ χρόνο,  η $f$ να περιέχει θόρυβο έτσι ώστε οι μέθοδοι που βασίζονται σε απειροστικές διαφορές να μην είναι αποτελεσματικές ή να υπάρχουν πολλά τοπικά βέλτιστα μεταξύ άλλων. 

Για περισσότερα πάνω στη βελτιστοποίηση χωρίς παραγώγους και τις εφαρμογές της, μπορείτε να συμβουλευτείτε το ["Gradient Free Optimization"](http://adl.stanford.edu/aa222/lecture_notes_files/chapter6_gradfree.pdf). 

Οφείλουμε εδώ να αναφέρουμε ότι κάποιες από τις συναρτήσεις που θα μελετήσουμε με ΓΑ μπορούν να βελτιστοποιηθούν πολύ ικανοποιητικά, ειδικά σε μικρές διαστάσεις, και με κλασικές αριθμητικές μεθόδους.

## Συναρτήσεις προς βελτιστοποίηση

Σε κάθε ομάδα αντιστοιχεί μία μη-κλιμακούμενη συνάρτηση (non-scalable, δηλαδή ορίζεται για συγκεκριμένες διαστάσεις του πεδίου ορισμού) και μία κλιμακούμενη (scalable, μπορεί να οριστεί για οποιαδήποτε διάσταση D). 

Καμία ομάδα δεν έχει τις ίδιες συναρτήσεις συνολικά με κάποια άλλη.

 Στο [“Ομάδες - Συναρτήσεις”](https://docs.google.com/spreadsheets/d/1g9otvvm1GIALF5c3Fsf45tULdoDZAmXnjmXyuW4sbro/edit?usp=sharing) θα βρείτε έναν αριθμό για τη μη-κλιμακούμενη συνάρτηση και έναν αριθμό για την κλιμακούμενη συνάρτηση που αντιστοιχούν στην ομάδα σας.
 
Θα βρείτε σε ποια συνάρτηση αντιστοιχεί ο κάθε αριθμός στο paper [“Α Literature Survey of Benchmark Functions For Global Optimization Problems”](https://goo.gl/qAhxNf), στο Section 3 (σελ. 5 και εξής). 

**Προσοχή:** ορισμένες συναρτήσεις όπως αναγράφονται στο paper έχουν μικρά (εώς μεσαία) λάθη ή χρειάζονται κάποια επεξήγηση. Έχουμε συγκεντρώσει τα errata που έχουμε ανακαλύψει μέχρι τώρα στο έγγραφο ["Errata στους ορισμούς συναρτήσεων"](https://drive.google.com/open?id=1MrRpq8q_hDfBZVw1CHBzC7t4biQkAftu3LEWXJMf6s0), παρακαλούμε συμβουλευτείτε το **απαραίτητα** για να δείτε αν αφορά σε κάποια από τις συναρτήσεις σας.

Οι ομάδες που έχουν μη-κλιμακούμενες συναρτήσεις με κωδικούς Α01-Α10 ή κλιμακούμενες συναρτήσεις με κωδικούς B01-B10 θα βρουν τους ορισμούς που τους αντιστοιχούν στο συμπληρωματικό έγγραφο ["Πρόσθετοι ορισμοί συναρτήσεων"](https://docs.google.com/document/d/1F-YJ5Ripo0N5ZAXCS52eGqNVB9RF6ZTvjnOh5ICHfss/edit?usp=sharing).

Όλες οι συναρτήσεις είναι προς **ελαχιστοποίηση**.


## Επιλογή κατάλληλου Γενετικού Αλγόριθμου

Ο κάθε Γενετικός Αλγόριθμος χαρακτηρίζεται α) από τους γενετικούς τελεστές του δηλαδή τον ορισμό της διασταύρωσης, μετάλλαξης και επιλογής και β) από τη στρατηγική εξέλιξης που καθορίζει την εξέλιξη του πληθυσμού μέσω της διαδικασίας επιλογής, καθώς και των  πιθανοτήτων διασταύρωσης και μετάλλαξης.

Όπως πολύ συχνά ισχύει σε προβλήματα Μηχανικής Μάθησης, η επιλογή του κατάλληλου αλγόριθμου εξαρτάται από το κάθε πρόβλημα και τα δεδομένα του. Συνεπώς για κάθε συνάρτηση θα πρέπει να δοκιμαστούν διαφορετικοί συνδυασμοί τελεστών και στρατηγικών.

Θα χρησιμοποιήσουμε ως γονίδια πραγματικούς αριθμούς και ως χρωμοσώματα λίστες γονιδίων μήκους ίσο με τη διάσταση του πεδίου ορισμού της συνάρτησης.



### Αξιολόγηση Γενετικών Αλγόριθμων

Μπορούμε να πούμε ότι ένας ΓΑ είναι πιθανόν να έχει καλύτερα χαρακτηριστικά για ένα δεδομένο πρόβλημα αν:
- βρίσκει τη λύση (βέλτιστη τιμή) με μεγαλύτερη συχνότητα
- βρίσκει τη βέλτιστη λύση με μικρότερο αριθμό αποτιμήσεων της συνάρτησης καταλληλότητας




#### Απόλυτα κριτήρια
Πρακτικά, για να αξιολογήσουμε τους αλγόριθμους της άσκησης θα φτιάξουμε μια συνάρτηση που θα αναλύει τα logbooks του αλγόριθμου ως εξής:  

Για ένα μέγιστο αριθμό γύρων (MAX_ROUNDS)  
&nbsp;&nbsp;&nbsp;&nbsp;Για ένα μέγιστο αριθμό γενεών (MAX_GENS)  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Για κάθε ΓΑ με δεδομένους γενετικούς τελεστές και στρατηγική εξέλιξης υπολογίζουμε τα εξής μεγέθη:

 
- **avg.min:** ο μέσος όρος της ελάχιστης τιμής που πέτυχε για όλους τους γύρους. Επί της ουσίας αυτό είναι το πιο σημαντικό και απόλυτο κριτήριο της επίδοσης του αλγόριθμου.
- **avg.evals:** o μέσος όρος των συνολικών αποτιμήσεων (άθροιση όλων των γενεών) που έκανε για όλους τους γύρους.
- **avg.time:** ο μέσος χρόνος εκτέλεσης των γύρων. 


#### Σχετικά κριτήρια

Τα προηγούμενα κριτήρια αποτελούν τα απόλυτα μέτρα της επίδοσης του αλγόριθμου.  Ωστόσο, επειδή δεν είναι εφικτό για κάθε συνάρτηση να αναμένουμε ότι σε λογικά χρονικά διαστήματα (ή και γενικά) θα βρούμε την ακριβή βέλτιστη λύση, ορίζουμε ως επιτυχία την ποσότητα **GOAL + DELTA**, όπου GOAL το ελάχιστο της συνάρτησης (δίνεται στο pdf ή στον ορισμό της συναρτησής σας) και DELTA μια θετική ποσότητα που θα ορίζουμε εμείς κάθε φορά (διαφορετική για κάθε συνάρτηση και διαφορετική ανάλογα του πόσο κοντά στη λύση θέλουμε να φτάσουμε). Με βάση την ποσότητα GOAL + DELTA ορίζουμε επιπλέον τα εξής μεγέθη:
- **successes:** αριθμός φορών που o ΓΑ βρήκε τιμή μικρότερη του GOAL + DELTA. Προφανώς το βέλτιστο είναι MAX_ROUNDS.
- **success avg. gen:** o μέσος όρος των γενεών που χρειάστηκε ο αλγόριθμος για να πέσει κάτω από GOAL + DELTA, για τους γύρους με επιτυχία. Ουσιαστικά μας ενδιαφέρει η πρώτη γενιά που πετυχαίνει τιμή κάτω από GOAL + DELTA. Σε περίπτωση που δεν πετυχαίνει αυτό το στόχο καμία γενιά, θέτουμε ‘None’ σε αυτή την ποσότητα.
- **success avg. min:** ο μέσος όρος των ελαχίστων των πρώτων επιτυχημένων γενεών για τους γύρους με επιτυχία. Παρόμοια, θέτουμε ‘None’ σε περίπτωση αποτυχίας.
- **success avg. evals:** ο μέσος όρος αποτιμήσεων που χρειάστηκαν για να πάρουμε την πρώτη τιμή κάτω από GOAL + DELTA. Και πάλι  θέτουμε ‘None’ σε περίπτωση αποτυχίας.

Πρόκειται ουσιαστικά για ένα σύνολο πιο "χαλαρών" κριτηρίων βελτιστοποίησης (λόγω του DELTA) που όμως θα μας βοηθήσουν να προσδιορίσουμε ευκολότερα τις υπερπαραμέτρους του τελικού γενετικού αλγόριθμου.




### Βοηθητική συνάρτηση εκτύπωσης υπερπαραμέτρων και απόδοσης (σχετικής και απόλυτης)

Φτιάχνουμε τέλος μια βοηθητική συνάρτηση που τυπώνει συγκεντρωτικά τα αποτελέσματα ως εξής (η επεξήγηση για τις δύο πρώτες κολόνες στο ακριβώς επόμενο section):

| operators                      | strategy                      | successes | s.avg.min | s.avg.evals | s.avg.gens | avg.evals | avg.min   | avg.time |
|--------------------------------|-------------------------------|-----------|-----------|-------------|------------|-----------|-----------|----------|
|CrossoverLow,MutationLow,SelectionLow | eaSimple 200 0.8 0.2          | 1         | 8.508e-42 | 16487       | 96         | 17014     | 9.115e-34 | 1.898    |
| CrossoverHigh,MutationLow,SelectionLow | eaMuPlusLambda 50 150 0.8 0.2 | 1         | 3.885e-41 | 14750       | 98         | 15050     | 1.631e-21 | 1.591    



## Μέρος 1. Βελτιστοποίηση μη κλιμακούμενης συνάρτησης



### Εύρεση βέλτιστου συνδυασμού τελεστών - στρατηγικής



#### Γενετικοί τελεστές

Αρχικά θα δοκιμάσουμε δύο διαφορετικούς τελεστές διασταύρωσης και δύο διαφορετικούς τελεστές μετάλλαξης της επιλογής μας, με δύο διαφορετικές τιμές υπερ-παραμέτρων για τον καθένα (μία χαμηλή-low και μία υψηλή-high - εάν κάποιος τελεστής έχει περισσότερες υπερπαραμέτρους απλά θεωρήστε δύο συνδυασμούς των τιμών τους, ο συνδυασμός high να οδηγεί σε μεγαλύτερη αλλαγή του γενετικού υλικού και αντίστροφα για τον συνδυασμό low). Θα δοκιμάσουμε επίσης δύο τιμές υπερ-παραμέτρων για τον τελεστή επιλογής selTournament. Έχουμε δηλαδή συνολικά 32 συνδυασμούς τελεστών. Σε αυτά αντιστοιχούν οι dummy ονομασίες της κολόνας "operators" (πλην του selTournament). Εσείς θα βάλετε τα κανονικά ονόματα των τελεστών με Low και High στο τέλος.



#### Στρατηγική εξέλιξης

Για κάθε συνδυασμό τελεστών θα εξετάσουμε τρεις στρατηγικές: τον **απλό ΓΑ** και τις στρατηγικές εξέλιξης **“μ+λ”** και **“μ,λ”**. Στην κολόνα strategy βλέπετε το όνομα της στρατηγικής και μετά την τιμή ή τις τιμές που χαρακτηρίζουν τον πληθυσμό. Προφανώς o απλός γενετικός έχει μόνο το μέγεθος του πληθυσμού ενώ οι δύο άλλες στρατηγικές έχουν τις τιμές για τα μ και λ. Οι δύο τελευταίοι αριθμοί της κολόνας αντιστοιχούν στις πιθανότητες διασταύρωσης και μετάλλαξης.



#### Μεθοδολογία εύρεσης βέλτιστου συνδυασμού

Αρχικά θα προσδιορίσουμε τους καταλληλότερους συνδυασμούς τελεστών-στρατηγικών. Για το λόγο αυτό θα θέσουμε ένα σχετικά μικρό αριθμό γύρων (π.χ. 5-10) και μέγιστων γενεών (π.χ. 50-100-150) και σταθερές τιμές στις παραμέτρους πληθυσμού και τις πιθανότητες διασταύρωσης και μετάλλαξης.  Διαλέξτε τιμές πληθυσμού έτσι ώστε οι στρατηγικές να έχουν συγκρίσιμο μέσο αριθμό αποτιμήσεων, να μπορούν να σημειωθούν επιτυχείς γύροι και κάθε γύρος να απαιτεί 1.5 - 2 δευτερόλεπτα. Διαλέξτε “λογικές” (μη ακραίες τιμές) για τις πιθανότητες διασταύρωσης και μετάλλαξης (στη βιβλιογραφία επιλέγεται συνήθως μεγαλύτερη πιθανότητα διασταύρωσης και μικρότερη μετάλλαξης).

Διαλέξτε ένα αρχικό DELTA κοντά στο ολικό ελάχιστο του καλύτερου συνδυασμού (από τα πρώτα runs που θα κάνετε). Το “κοντά” σε αυτό το στάδιο μπορούμε να το εκτιμήσουμε μόνο εμπειρικά: επαναλάβετε με διάφορα DELTA έτσι ώστε στην τελική επιλογή DELTA οι καλύτεροι συνδυασμοί να έχουν ποσοστό επιτυχιών κοντά στη μονάδα και οι χειρότεροι καμία. 



In [154]:
#!pip install deap

from deap import base, creator, tools
import multiprocessing
import random

creator.create("FitnessMin", base.Fitness, weights=(-1.0,))
creator.create("Individual", list, fitness=creator.FitnessMin)

toolbox = base.Toolbox()
toolbox.register("attr", random.uniform, -4.5, 4.5)
toolbox.register("individual", tools.initRepeat, creator.Individual, toolbox.attr, 2)
toolbox.register("population", tools.initRepeat, list, toolbox.individual)

pool = multiprocessing.Pool()
toolbox.register("map", pool.map)

/usr/local/lib/python3.6/dist-packages/deap/creator.py:141: RuntimeWarning: A class named 'FitnessMin' has already been created and it will be overwritten. Consider deleting previous creation of that class or rename it.
  RuntimeWarning)
/usr/local/lib/python3.6/dist-packages/deap/creator.py:141: RuntimeWarning: A class named 'Individual' has already been created and it will be overwritten. Consider deleting previous creation of that class or rename it.
  RuntimeWarning)


In [0]:
def evalBeale(x):
  x1, x2 = x
  if(abs(x1) > 4.5 or abs(x2) > 4.5):
    return 1000.0,
  return (1.5 - x1 +x1*x2)**2 + (2.25 - x1 + x1*x2**2)**2 + (2.65 - x1 + x1*x2**3)**2,

In [0]:
toolbox.register("evaluate", evalBeale)

In [0]:
from itertools import product
from deap import algorithms

In [0]:
MAX_ROUNDS = 15
MAX_GENS = 100
POP_SIZE = 100
DELTA = 0.0001

mutation = ['mutGaussianHigh', 'mutGaussianLow', 'mutPolyHigh', 'mutPolyLow']
crossover = ['crossBlendHigh', 'crossBlendLow', 'crossSimBinHigh', 'crossSimBinLow']
select = ['selTournamentHigh','selTournamentLow']
algorithm = ['SimpleEA', 'MuCommaLambda', 'MuPlusLambda']

mut_parameters = {
    'mutGaussianHigh': (tools.mutGaussian, {'mu':0, 'sigma':0.45, 'indpb':0.75}),
    'mutGaussianLow': (tools.mutGaussian, {'mu':0, 'sigma':0.09, 'indpb':0.25}), 
    'mutPolyHigh': (tools.mutPolynomialBounded, {'eta':4, 'low':-4.5, 'up':4.5, 'indpb':0.75}),  # low eta -> high mutation
    'mutPolyLow': (tools.mutPolynomialBounded, {'eta':50, 'low':-4.5, 'up':4.5, 'indpb':0.25})
}

cross_parameters = {
    'crossBlendHigh':(tools.cxBlend, {'alpha':0.2}),
    'crossBlendLow':(tools.cxBlend, {'alpha':0.05}),
    'crossSimBinHigh':(tools.cxSimulatedBinaryBounded, {'eta':4, 'low':-4.5, 'up':4.5}), # low eta -> high crossover
    'crossSimBinLow':(tools.cxSimulatedBinaryBounded, {'eta':50, 'low':-4.5, 'up':4.5})
}

select_parameters = {
    'selTournamentHigh':(tools.selTournament, {'tournsize':10}),
    'selTournamentLow':(tools.selTournament, {'tournsize':3})
}

algo = {
    'SimpleEA': algorithms.eaSimple, 
    'MuCommaLambda': algorithms.eaMuCommaLambda, 
    'MuPlusLambda': algorithms.eaMuPlusLambda
}

algo_parameters = {
    'SimpleEA': (algorithms.eaSimple, {'cxpb':0.5, 'mutpb':0.2}), 
    'MuCommaLambda': (algorithms.eaMuCommaLambda, {'mu':POP_SIZE, 'lambda_':int(POP_SIZE*1.25), 'cxpb':0.5, 'mutpb':0.2}), 
    'MuPlusLambda': (algorithms.eaMuPlusLambda, {'mu':POP_SIZE, 'lambda_':int(POP_SIZE*0.25), 'cxpb':0.5, 'mutpb':0.2})
}

In [0]:
import time
import pandas as pd
import numpy as np

if __name__ == "__main__":
  i = 0
  df = pd.DataFrame(columns=['operators', 'strategy', 'successes', 's.avg.min',\
               's.avg.evals', 's.avg.gens', 'avg.evals', 'avg.min', 'avg.time'])

  for parameters in product(algorithm, crossover, mutation, select):
    
    min_sum = 0
    eval_sum = 0
    time_sum = 0
    
    successes = 0
    gen_sum = 0
    suc_min_sum = 0
    suc_evals = 0   

    algo_, cross_, mut_, select_ = parameters
    operators = ', '.join([cross_, mut_, select_])
    strategy = algo_
    # mutate
    mut_, mut_params = mut_parameters[mut_]
    toolbox.register("mutate", mut_, **mut_params)
    # mate
    cross_, cross_params = cross_parameters[cross_]
    toolbox.register("mate", cross_, **cross_params)
    # select
    select_, select_params = select_parameters[select_]
    toolbox.register("select", select_, **select_params)
    # algorithm
    algo_2, algo_params = algo_parameters[algo_]
    
    for r in range(MAX_ROUNDS):
      
      pop = toolbox.population(n=POP_SIZE)
      hof = tools.HallOfFame(1)
      stats = tools.Statistics(lambda ind: ind.fitness.values)
      stats.register("min", np.min)
      
      t0 = time.time()
      pop, logbook = algo[algo_](population=pop, toolbox=toolbox, halloffame=hof,\
                      stats=stats, ngen=MAX_GENS, verbose=False, **algo_params)
      t1 = time.time()
      
      #calculate stats
      min_sum += hof[0].fitness.values[0]
      eval_sum += sum(logbook.select('nevals'))
      time_sum += t1 - t0
      successes += hof[0].fitness.values[0] < DELTA
      gen_success = np.array(logbook.select('min')) < DELTA
      if any(gen_success):
        first_gen = np.argmax(gen_success)
        gen_sum += first_gen
        suc_min_sum += logbook.select('min')[first_gen]
        suc_evals += sum(logbook.select('nevals')[:first_gen])
        
                           
    i += 1
    print('{}/96'.format(i))
    
    if(successes == 0):
      s_avg_min = None
      s_avg_evals = None
      s_avg_gens = None
    else:
      s_avg_min = suc_min_sum/successes
      s_avg_evals = suc_evals/successes
      s_avg_gens = gen_sum/successes
      
    df.loc[i] = ([operators, strategy, successes, s_avg_min, s_avg_evals, s_avg_gens,\
                  eval_sum/MAX_ROUNDS, min_sum/MAX_ROUNDS, time_sum/MAX_ROUNDS])

In [261]:
df.sort_values('avg.min', ascending=True)

operators       strategy  \
33  crossBlendHigh, mutGaussianHigh, selTournament...  MuCommaLambda   
41  crossBlendLow, mutGaussianHigh, selTournamentHigh  MuCommaLambda   
34  crossBlendHigh, mutGaussianHigh, selTournamentLow  MuCommaLambda   
36   crossBlendHigh, mutGaussianLow, selTournamentLow  MuCommaLambda   
10   crossBlendLow, mutGaussianHigh, selTournamentLow       SimpleEA   
1   crossBlendHigh, mutGaussianHigh, selTournament...       SimpleEA   
49  crossSimBinHigh, mutGaussianHigh, selTournamen...  MuCommaLambda   
11   crossBlendLow, mutGaussianLow, selTournamentHigh       SimpleEA   
35  crossBlendHigh, mutGaussianLow, selTournamentHigh  MuCommaLambda   
47       crossBlendLow, mutPolyLow, selTournamentHigh  MuCommaLambda   
2   crossBlendHigh, mutGaussianHigh, selTournamentLow       SimpleEA   
51  crossSimBinHigh, mutGaussianLow, selTournament...  MuCommaLambda   
50  crossSimBinHigh, mutGaussianHigh, selTournamen...  MuCommaLambda   
43   crossBlendLow, mutGaussianLow, selTournamentHigh  MuCommaLambda   
39      crossBlendHigh, mutPolyLow, selTournamentHigh  MuCommaLambda   
38      crossBlendHigh, mutPolyHigh, selTournamentLow  MuCommaLambda   
18  crossSimBinHigh, mutGaussianHigh, selTournamen...       SimpleEA   
58  crossSimBinLow, mutGaussianHigh, selTournamentLow  MuCommaLambda   
14       crossBlendLow, mutPolyHigh, selTournamentLow       SimpleEA   
37     crossBlendHigh, mutPolyHigh, selTournamentHigh  MuCommaLambda   
12    crossBlendLow, mutGaussianLow, selTournamentLow       SimpleEA   
59  crossSimBinLow, mutGaussianLow, selTournamentHigh  MuCommaLambda   
46       crossBlendLow, mutPolyHigh, selTournamentLow  MuCommaLambda   
16        crossBlendLow, mutPolyLow, selTournamentLow       SimpleEA   
48        crossBlendLow, mutPolyLow, selTournamentLow  MuCommaLambda   
8        crossBlendHigh, mutPolyLow, selTournamentLow       SimpleEA   
4    crossBlendHigh, mutGaussianLow, selTournamentLow       SimpleEA   
44    crossBlendLow, mutGaussianLow, selTournamentLow  MuCommaLambda   
56      crossSimBinHigh, mutPolyLow, selTournamentLow  MuCommaLambda   
57  crossSimBinLow, mutGaussianHigh, selTournament...  MuCommaLambda   
..                                                ...            ...   
60   crossSimBinLow, mutGaussianLow, selTournamentLow  MuCommaLambda   
89  crossSimBinLow, mutGaussianHigh, selTournament...   MuPlusLambda   
73  crossBlendLow, mutGaussianHigh, selTournamentHigh   MuPlusLambda   
32       crossSimBinLow, mutPolyLow, selTournamentLow       SimpleEA   
78       crossBlendLow, mutPolyHigh, selTournamentLow   MuPlusLambda   
20  crossSimBinHigh, mutGaussianLow, selTournamentLow       SimpleEA   
7       crossBlendHigh, mutPolyLow, selTournamentHigh       SimpleEA   
77      crossBlendLow, mutPolyHigh, selTournamentHigh   MuPlusLambda   
90  crossSimBinLow, mutGaussianHigh, selTournamentLow   MuPlusLambda   
71      crossBlendHigh, mutPolyLow, selTournamentHigh   MuPlusLambda   
88      crossSimBinHigh, mutPolyLow, selTournamentLow   MuPlusLambda   
3   crossBlendHigh, mutGaussianLow, selTournamentHigh       SimpleEA   
93     crossSimBinLow, mutPolyHigh, selTournamentHigh   MuPlusLambda   
72       crossBlendHigh, mutPolyLow, selTournamentLow   MuPlusLambda   
84  crossSimBinHigh, mutGaussianLow, selTournamentLow   MuPlusLambda   
81  crossSimBinHigh, mutGaussianHigh, selTournamen...   MuPlusLambda   
67  crossBlendHigh, mutGaussianLow, selTournamentHigh   MuPlusLambda   
79       crossBlendLow, mutPolyLow, selTournamentHigh   MuPlusLambda   
28   crossSimBinLow, mutGaussianLow, selTournamentLow       SimpleEA   
95      crossSimBinLow, mutPolyLow, selTournamentHigh   MuPlusLambda   
85    crossSimBinHigh, mutPolyHigh, selTournamentHigh   MuPlusLambda   
70      crossBlendHigh, mutPolyHigh, selTournamentLow   MuPlusLambda   
75   crossBlendLow, mutGaussianLow, selTournamentHigh   MuPlusLambda   
68   crossBlendHigh, mutGaussianLow, selTournamentLow   MuPlusLambda   
31      cros


#### Τελική αναζήτηση και σχολιασμός

Κρατήστε μόνο την τελική αναζήτηση σε μπλοκ κώδικα και περιγράψτε την πορεία των επιλογών σας σε markdown. 

Σχολιάστε συνολικά τα αποτελέσματα και προτείνετε τους καλύτερους συνδυασμούς τελεστών-στρατηγικών για τη συνάρτησή σας (αιτιολογήστε την επιλογή σας).


Ως καλύτερο συνδυασμό επιλέγουμε τον crossBlendHigh, mutGaussianHigh, selTournamentHigh,	MuCommaLambda γιατί έχει το βέλτιστο avg.min και πολύ καλές τιμές σε όλες τις υπόλοιπες μετρικές. 

Έχει 10 successes που είναι το 2ο καλύτερο μετά το 11 και επομένως δινεί διαρκώς καλά αποτελέσματα.  
Έχει 8.6 s.avg.evals που είναι το 2ο καλύτερο μετά το 8.5 και επομένως δίνει με πολύ λίγες γενίες καλά αποτελέσματα. 


### Τελική βελτιστοποίηση



#### Βελτιστοποίηση πιθανοτήτων διασταύρωσης και μετάλλαξης

Για τον καλύτερο συνδυασμό τελεστών-στρατηγικής (αιτιολογήστε την επιλογή σας) και σταθερό αριθμό γύρων, πληθυσμού και γενεών, κάντε grid search στις πιθανότητες των τελεστών διασταύρωσης και μετάλλαξης στο διάστημα [0.05 0.9] ώστε να βρείτε το βέλτιστο συνδυασμό τους. 

Μπορείτε να δοκιμάσετε σταδιακά πιο μικρά διαστήματα περιοχών τιμών (αύξηση της ακρίβειας) γύρω από καλές τιμές.


In [0]:
toolbox.register("mutate", tools.mutGaussian, mu=0, sigma=0.45, indpb=0.75)
toolbox.register("mate", tools.cxBlend, alpha=0.2)
toolbox.register("select", tools.selTournament, tournsize=10)
algorithm = algorithms.eaMuPlusLambda

Χώροι αναζήτησης:  
Προσοχή πάντα ώστε max(cxpb) + max(mutpb) <= 1
1. cxpb_ = np.linspace(0.05, 0.5, 6)  
mutpb_ = np.linspace(0.05, 0.5, 6)  

2. cxpb_ = np.linspace(0.5, 0.9, 5)  
mutpb_ = np.linspace(0.05, 0.1, 2)

3. cxpb_ = np.linspace(0.05, 0.1, 2)  
mutpb_ = np.linspace(0.5, 0.9, 5)  

Μέχρι στιγμής όσο πιο υψηλές τιμές τόσο καλύτερα αποτελέσματα. Γενικά μεγάλο mutpb -> μικρό min, μικρό successes. Μεγάλο cxpb -> μεγάλο (μεγαλύτερο) min, μεγάλο successes. 

4. cxpb_ = np.linspace(0.05, 0.1, 6)  
mutpb_ = np.linspace(0.9, 0.8, 11)

5. cxpb_ = np.linspace(0.9, 0.8, 11)  
mutpb_ = np.linspace(0.05, 0.1, 6)

cxpb = 0.87 και mutpb =	0.08
έχει πολύ καλό avg.min και εξαιρετικά καλό αριθμό successes (14/15)

In [0]:
cxpb_ = np.linspace(0.9, 0.8, 11)
mutpb_ = np.linspace(0.05, 0.1, 6)

In [320]:
MAX_ROUNDS = 15
MAX_GENS = 100
POP_SIZE = 100
DELTA = 0.0001

if __name__ == "__main__":
  i = 0
  df = pd.DataFrame(columns=['cxpb', 'mutpb', 'successes', 's.avg.min',\
               's.avg.evals', 's.avg.gens', 'avg.evals', 'avg.min', 'avg.time'])

  for cxpb, mutpb in product(cxpb_, mutpb_):
    
    min_sum = 0
    eval_sum = 0
    time_sum = 0
    
    successes = 0
    gen_sum = 0
    suc_min_sum = 0
    suc_evals = 0   

    for r in range(MAX_ROUNDS):
      
      pop = toolbox.population(n=POP_SIZE)
      hof = tools.HallOfFame(1)
      stats = tools.Statistics(lambda ind: ind.fitness.values)
      stats.register("min", np.min)
      
      t0 = time.time()
      pop, logbook = algorithm(population=pop, toolbox=toolbox, halloffame=hof,\
                      stats=stats, ngen=MAX_GENS, verbose=False, mu=POP_SIZE,\
                      lambda_=int(POP_SIZE*1.25), cxpb=cxpb, mutpb=mutpb)
      t1 = time.time()
      
      #calculate stats
      min_sum += hof[0].fitness.values[0]
      eval_sum += sum(logbook.select('nevals'))
      time_sum += t1 - t0
      successes += hof[0].fitness.values[0] < DELTA
      gen_success = np.array(logbook.select('min')) < DELTA
      if any(gen_success):
        first_gen = np.argmax(gen_success)
        gen_sum += first_gen
        suc_min_sum += logbook.select('min')[first_gen]
        suc_evals += sum(logbook.select('nevals')[:first_gen])
        
                           
    i += 1
    print('{}/66'.format(i))
    
    if(successes == 0):
      s_avg_min = None
      s_avg_evals = None
      s_avg_gens = None
    else:
      s_avg_min = suc_min_sum/successes
      s_avg_evals = suc_evals/successes
      s_avg_gens = gen_sum/successes
      
    df.loc[i] = ([cxpb, mutpb, successes, s_avg_min, s_avg_evals, s_avg_gens,\
                  eval_sum/MAX_ROUNDS, min_sum/MAX_ROUNDS, time_sum/MAX_ROUNDS])
  

1/66
2/66
3/66
4/66
5/66
6/66
7/66
8/66
9/66
10/66
11/66
12/66
13/66
14/66
15/66
16/66
17/66
18/66
19/66
20/66
21/66
22/66
23/66
24/66
25/66
26/66
27/66
28/66
29/66
30/66
31/66
32/66
33/66
34/66
35/66
36/66
37/66
38/66
39/66
40/66
41/66
42/66
43/66
44/66
45/66
46/66
47/66
48/66
49/66
50/66
51/66
52/66
53/66
54/66
55/66
56/66
57/66
58/66
59/66
60/66
61/66
62/66
63/66
64/66
65/66
66/66


In [321]:
df.sort_values('avg.min', ascending=True)

cxpb  mutpb  successes  s.avg.min  s.avg.evals  s.avg.gens     avg.evals  \
22  0.87   0.08       14.0   0.000097  1120.142857    9.571429  11967.066667   
46  0.83   0.08       13.0   0.000096  1186.769231   10.538462  11469.200000   
18  0.88   0.10       13.0   0.000096  1333.538462   11.076923  12346.600000   
66  0.80   0.10       13.0   0.000096  1746.769231   15.615385  11359.133333   
9   0.89   0.07       13.0   0.000096  1455.923077   12.307692  12098.200000   
24  0.87   0.10       12.0   0.000096   930.916667    7.833333  12217.933333   
40  0.84   0.08       13.0   0.000097  1321.692308   11.615385  11598.066667   
60  0.81   0.10       12.0   0.000096   906.916667    8.083333  11480.266667   
14  0.88   0.06       12.0   0.000096  1482.000000   12.750000  11845.866667   
1   0.90   0.05       10.0   0.000097  1098.100000    9.400000  11974.600000   
59  0.81   0.09       12.0   0.000097  1337.500000   12.000000  11345.533333   
38  0.84   0.06       10.0   0.000096  1880.300000   16.800000  11356.066667   
25  0.86   0.05       12.0   0.000096   952.333333    8.500000  11478.066667   
5   0.90   0.09       10.0   0.000096  2771.600000   22.600000  12475.666667   
65  0.80   0.09       10.0   0.000097   859.200000    7.800000  11238.133333   
15  0.88   0.07       12.0   0.000096  1415.833333   12.083333  11969.600000   
64  0.80   0.08       11.0   0.000097   891.454545    8.181818  11094.800000   
23  0.87   0.09       13.0   0.000097  2420.692308   20.307692  12110.866667   
48  0.83   0.10       10.0   0.000096  2154.800000   18.700000  11717.933333   
49  0.82   0.05       11.0   0.000096  1114.090909   10.272727  10983.600000   
62  0.80   0.06       11.0   0.000096  1566.272727   14.636364  10853.533333   
10  0.89   0.08        9.0   0.000096   935.555556    7.888889  12218.866667   
36  0.85   0.10        9.0   0.000096  1067.111111    9.111111  11973.333333   
27  0.86   0.07       10.0   0.000097  1501.900000   13.000000  11730.533333   
56  0.81   0.06       10.0   0.000097  1779.000000   16.400000  10978.266667   
8   0.89   0.06       13.0   0.000096  1214.076923   10.384615  11970.266667   
30  0.86   0.10        9.0   0.000097   977.222222    8.333333  12098.400000   
20  0.87   0.06       10.0   0.000096   971.400000    8.500000  11723.266667   
28  0.86   0.08       10.0   0.000098  2939.500000   25.200000  11845.800000   
21  0.87   0.07       11.0   0.000096  1629.818182   14.000000  11857.400000   
..   ...    ...        ...        ...          ...         ...           ...   
41  0.84   0.09       10.0   0.000097   903.300000    7.900000  11739.800000   
12  0.89   0.10       13.0   0.000097  1584.692308   13.000000  12473.266667   
31  0.85   0.05       13.0   0.000096  1519.692308   13.615385  11362.466667   
45  0.83   0.07        9.0   0.000096  3446.777778   30.666667  11352.000000   
6   0.90   0.10       13.0   0.000096  1292.307692   10.538462  12600.000000   
55  0.81   0.05        7.0   0.000097  1894.714286   17.714286  10834.866667   
2   0.90   0.06       12.0   0.000097  2727.916667   22.916667  12102.066667   
7   0.89   0.05        7.0   0.000096   972.571429    8.428571  11851.933333   
34  0.85   0.08        7.0   0.000096  1028.571429    9.000000  11723.000000   
47  0.83   0.09       10.0   0.000096  1945.900000   17.000000  11592.133333   
54  0.82   0.10        8.0   0.000097  1726.250000   15.125000  11599.800000   
53  0.82   0.09        9.0   0.000097  1765.444444   15.666667  11464.400000   
42  0.84   0.10        9.0   0.000096  1455.222222   12.555556  11846.800000   
33  0.85   0.07        9.0   0.000096  1136.444444   10.000000  11607.933333   
11  0.89   0.09        9.0   0.000097   836.888889    7.000000  12356.200000   
57  0.81   0.07        9.0   0.000096  2461.000000   22.444444  11102.000000   
26  0.86   0.06        8.0   0.000097  1045.375000    9.250000  11608.000000   
4   0.90   0.08        8.0   0.000097  1277.625000   10.625000  12350.200000   
13  


#### Εύρεση βέλτιστης (ελάχιστης) τιμής της συνάρτησης με τον ΓΑ

Κάντε ένα run του βέλτιστου αλγόριθμου που προέκυψε στα προηγούμενα βήματα με ένα μεγάλο αριθμό γενεών και πληθυσμού για να πάρετε μια βέλτιστη τιμή για τη συνάρτηση (σε λογικά χρονικά πλαίσια). Τυπώστε τη βέλτιστη τιμή, το συνολικό αριθμό αποτιμήσεων και το χρόνο εκτέλεσης.


In [327]:
POP_SIZE = 20000
MAX_GENS = 20

if __name__ == "__main__":
  pop = toolbox.population(n=POP_SIZE)
  hof = tools.HallOfFame(1)
  stats = tools.Statistics(lambda ind: ind.fitness.values)
  stats.register("min", np.min)

  t0 = time.time()
  pop, logbook = algorithm(population=pop, toolbox=toolbox, halloffame=hof,\
                  stats=stats, ngen=MAX_GENS, verbose=True, mu=POP_SIZE,\
                  lambda_=int(POP_SIZE*1.25), cxpb = 0.87, mutpb =	0.08)
  t1 = time.time()
  
  neval = sum(logbook.select('nevals'))
  time_delta = t1 - t0
  gen_success = np.array(logbook.select('min')) < DELTA
  first_gen = np.argmax(gen_success)
  suc_evals = sum(logbook.select('nevals')[:first_gen])

gen	nevals	min       
0  	20000 	0.00370711
1  	23777 	0.00045965
2  	23773 	0.00018781
3  	23742 	0.000117394
4  	23755 	9.78688e-05
5  	23771 	9.78688e-05
6  	23788 	9.45633e-05
7  	23720 	9.41454e-05
8  	23791 	9.41394e-05
9  	23735 	9.41385e-05
10 	23742 	9.41384e-05
11 	23707 	9.41384e-05
12 	23750 	9.41384e-05
13 	23810 	9.41384e-05
14 	23707 	9.41384e-05
15 	23811 	9.41384e-05
16 	23751 	9.41384e-05
17 	23742 	9.41384e-05
18 	23691 	9.41384e-05
19 	23701 	9.41384e-05
20 	23764 	9.41384e-05


In [328]:
print(hof[0])
print(time_delta)
print(first_gen)
print(suc_evals)

[3.0505948817753903, 0.5098630541476916]
47.42425775527954
4
91292


Το πραγματικό βέλτιστο είναι το 0 στο (3, 0.5). Ο αλγόριθμος βρήκε μία εξαιρετικά καλή προσέγγιση της λύσης και της βέλτιστης τιμής. Έπεσε μάλιστα κάτω από το 0.0001 σε μόλις 4 γενιές.


## Μέρος 2. Μελέτη κλιμακούμενης συνάρτησης

Όπως είδαμε και στο παράδειγμα του 0-1 Knapsack, πολλά προβλήματα μπορεί να λύνονται εύκολα σε μικρές διαστάσεις αλλά γίνονται δυσκολότερα όσο οι διαστάσεις μεγαλώνουν. Συνεπώς, μια επιθυμητή ιδιότητα για έναν αλγόριθμο είναι να μπορεί να αντιμετωπίζει την κλιμάκωση των προβλημάτων. Συνεπώς, η μελέτη της κλιμακούμενης συνάρτησή σας συνίσταται στα ακόλουθα ζητούμενα.



### Για D=2

- α) Εκτυπώστε ένα “3D” γράφημα της συνάρτησης $f(x1,x2)$ και περιγράψτε σύντομα τη μορφή της (βέλτιστα, κοίλα, κοιλάδες, λεκάνες, ασυνέχειες κλπ)
- β) Με την διαδικασία που ακολουθήσαμε προηγουμένως για τη μη κλιμακούμενη συνάρτηση, βρείτε τον βέλτιστο γενετικό αλγόριθμο και τη βέλτιστη τιμή για το πρόβλημα.



### Για D=1, 10, 20, 40 και μεγαλύτερες διαστάσεις

Για τον βέλτιστο αλγόριθμο (που βρήκατε για τις 2 διαστάσεις) και για διαφορετικές τιμές/τάξεις μεγέθους D=1, 10, 20, 40 ή και περισσότερων διαστάσεων του πεδίου ορισμού (σταθερά MAX_GENS και MAX_ROUNDS>=10) τυπώστε πίνακα με: **αριθμό διαστάσεων**, **αριθμό επιτυχιών**, **μέσο ολικό ελάχιστο**, **μέσο αριθμό αποτιμήσεων** και **μέσο χρόνο** (προφανώς βασιστείτε στην αρχική βοηθητική συνάρτηση εκτύπωσης αποτελεσμάτων). 

Σημειώστε ότι οποιεσδήποτε ποινές για άτομα εκτός ορισμού καθώς και τo DELTA, προκειμένου να είναι αποτελεσματικά ως προς την κλιμάκωση των διαστάσεων, θα πρέπει να οριστούν εμπειρικά μεν, αλλά οπωσδήποτε συναρτήσει του πλήθους των διαστάσεων (μεγαλύτερες ποινές για μεγαλύτερες διαστάσεις, για το DELTA έχουν δοθεί οδηγίες σε προηγούμενο section).

α) Σχολιάστε τα αποτελέσματα του πίνακα.

β) Ποιες πιστεύετε είναι οι αιτίες του φαινομένου;

γ) Για σταθερό αριθμό γενεών, πώς μπορούμε να βελτιώσουμε τα αποτελέσματα σε μεγάλες διαστάσεις ως προς το σκέλος της μέσης βέλτιστης τιμής; (Mπορεί να υπάρχουν περισσότερες προσεγγίσεις στο ίδιο πρόβλημα.)

**Σημείωση:** όπως συνήθως συμβαίνει στα υπολογιστικά προβλήματα, μεγαλύτερες διαστάσεις εισόδου σημαίνουν και μεγαλύτερους χρόνους εκτέλεσης. Ταυτόχρονα, οι προς μελέτη συναρτήσεις όλων των ομάδων έχουν διαφορετικές ιδιότητες και βαθμούς δυσκολίας. Συνεπώς, είναι αδύνατον να δώσουμε γενική οδηγία του τί είναι μεγάλες ή μεγαλύτερες διαστάσεις που να ισχύει για όλες τις συναρτήσεις. Δίνουμε κάποιες ενδεικτικές τιμές αλλά μπορείτε να τις προσαρμόσετε (και προς τα κάτω αλλά κυρίως προς τα πάνω, πχ αν οι χρόνοι είναι της τάξης λεπτών δοκιμάστε 100, 200, 500 κοκ) ανάλογα με τους διαθέσιμους πόρους, τον αριθμό επαναλήψεων και του τί θεωρείτε λογικά πλαίσια χρόνου αναμονής.


### Βελτιστοποίηση σε μεγάλες διαστάσεις

Με βάση τις απαντήσεις σας στο πρηγούμενο (β) και πάντοτε με σταθερό αριθμό γενεών, εφαρμόστε τις αλλαγές που χρειάζονται σύμφωνα με την απάντησή σας στο (γ) με τους εξής δύο στόχους:

1. Κρατώντας σταθερή και μεγάλη τη διάσταση εισόδου να παίρνετε όλο και καλύτερες βέλτιστες τιμές της συνάρτησης. Παρουσιάστε τα αποτελέσματα σε συγκριτικό πίνακα παρόμοιο με τον προηγούμενο.

2. Βρείτε μια νέα -σταθερή- διάσταση, πιθανότατα μικρότερη από την προηγούμενη και ένα DELTA που να σας δίνουν 35% - 50% επιτυχίες. Μπορείτε να μειώσετε το DELTA στο μισό και να διπλασιάσετε το ποσοστό των επιτυχιών; Μπορείτε να ξαναμειώσετε στο μισό το DELTA και να αυξήσετε επιπρόσθετα το ποσοστό των επιτυχιών; Αν όχι, κατα πόσο κατ'ελάχιστον πρέπει να αυξήσετε τον αριθμό των γενεών για να πετύχετε σημαντική αύξηση των επιτυχιών;



## Ευρετήρια συναρτήσεων για benchmarking στη βελτιστοποίηση

Μπορείτε να συμβουλευτείτε τους παρακάτω καταλόγους συναρτήσεων benchmarking για διευκρινίσεις σχετικά με τις συναρτήσεις σας.

- [Electric Power Systems Analysis & Nature-Inspired Optimization](https://al-roomi.org/benchmarks/unconstrained)

- [Global Optimization Test Functions Index](http://infinity77.net/global_optimization/test_functions.html)

- [BenchmarkFcns Toolbox](http://benchmarkfcns.xyz/fcns)

- [Virtual Library of Simulation Experiments: Test Functions and Datasets](https://www.sfu.ca/~ssurjano/optimization.html)

- [DEAP Benchmarks functions](http://deap.readthedocs.io/en/master/api/benchmarks.html)




## Βιβλιογραφία

- Jamil, Momin, and Xin-She Yang. "A literature survey of benchmark functions for global optimisation problems." International Journal of Mathematical Modelling and Numerical Optimisation 4.2 (2013): 150-194. [link](https://goo.gl/qAhxNf). Προσοχή στα [errata](https://docs.google.com/document/d/1MrRpq8q_hDfBZVw1CHBzC7t4biQkAftu3LEWXJMf6s0/edit?usp=sharing) στους ορισμούς των συναρτήσεων.

- Geitle, Marius. Improving differential evolution using inductive programming. MS thesis. 2017. [link](https://goo.gl/QnjGjU)

- Adorio, Ernesto P., and U. Diliman. "Mvf-multivariate test functions library in c for unconstrained global optimization." Quezon City, Metro Manila, Philippines (2005): 100-104. [link](https://goo.gl/zTgYZE)

- Whitley, Darrell, et al. "Evaluating evolutionary algorithms." Artificial intelligence 85.1-2 (1996): 245-276. [link](https://www.sciencedirect.com/science/article/pii/0004370295001247/pdf?md5=41088a944920336e5d8493160de27800&pid=1-s2.0-0004370295001247-main.pdf&_valck=1)

- "Gradient Free Optimization". Stanford AA222 - Introduction to Multidisciplinary Design Optimization handouts. [link](http://adl.stanford.edu/aa222/Lecture_Notes_files/chapter6_gradfree.pdf) 






## Σημείωσεις για την υλοποίηση

- Λόγω των πολλών επαναλήψεων είναι απαραίτητη και θα αξιολογηθεί η **εκτεταμένη χρήση συναρτήσεων**.
- Για απορίες και διευκρυνίσεις ως συνήθως συμβουλευτείτε πρώτα το [FAQ](https://docs.google.com/document/d/1jL4gRag_LHbVCYIt5XVJ53iJPb6RZWi02rT5mPXiqEU/edit?usp=sharing) και στη συνέχεια αν χρειαστεί μπορείτε να στείλετε mail στο [nnlab@islab.ntua.gr](mailto:nnlab@islab.ntua.gr)

## Παράδοση

### Ημερομηνία παράδοσης

Πέμπτη 28 Φεβρουαρίου 2019 (παρακαλούμε όχι αιτήματα για παράταση).

### Παραδοτέα

Στο mycourses του μαθήματος εντός ενός zip file:

- το αρχείο ipynb του notebook σας με τις απαντήσεις σας (ορατά τα κελιά εξόδου σύμφωνα με το FAQ). Μην ξεχάσετε ένα κελί με τα ονόματα σας καθώς και τον αριθμό της ομάδας σας.
- ο καθαρός κώδικας Python του notebook δηλαδή το αρχείο .py.


<table>
  <tr>
    <td bgcolor="#FCF8E3"><font size="4">Παρακαλούμε διατρέξτε βήμα-βήμα το notebook για να μην ξεχάσετε παραδοτέα!!</font>
</td>
  </tr>
</table>

**Καλή επιτυχία!** 👍